In [1]:
import pandas as pd
import numpy as np

In [8]:
test_set = pd.read_csv('data/eval_dataset1.csv')
test_set.head(10)

instance_id  KLIJENT_ID  OZNAKA_PARTIJE DATUM_OTVARANJA  \
0            0     1049000         5585453      14.07.2012   
1            1      481302          752335      24.01.2016   
2            2      838192         6623548      26.12.2018   
3            3      174575          425060      04.08.2012   
4            4       14884         1183149      27.03.2011   
5            5     1348702        12847410      04.08.2018   
6            6      764915         5647773      21.12.2013   
7            7      367476         4875789      08.01.2010   
8            8      234350        12636299      08.08.2017   
9            9      888283         9088014      23.11.2018   

  PLANIRANI_DATUM_ZATVARANJA  UGOVORENI_IZNOS  VALUTA  VRSTA_KLIJENTA  \
0                  7/15/2013        273555.01       2            1410   
1                  1/24/2016         50970.78       2            1410   
2                 12/26/2018          8969.36       2            1410   
3                   8/4/2012         16142.48       2            1410   
4                  3/28/2012         45315.31       2            1410   
5                  2/15/2020         40000.00       1            1410   
6                  3/22/2014        119043.94       2            1410   
7                   1/9/2011        103650.71       2            1410   
8                   8/9/2018        185000.00       1            1550   
9                 11/23/2018         54123.73       1            1410   

  PROIZVOD VRSTA_PROIZVODA  VISINA_KAMATE TIP_KAMATE  STAROST  \
0   TM0109               L           3.60          A       64   
1   TM0109               L           0.20          A       56   
2   TM0109               L           0.02          A       52   
3   TM0109               L           2.30          C       76   
4   TM0109               L           3.30          C       53   
5   TM0109               L           1.50          A       35   
6   TM0109               L           2.00          A       50   
7   TM0109               L           5.15          A       35   
8   TM0109               L           1.30          A       46   
9   TM0109               L           0.20          A       38   

   PRIJEVREMENI_RASKID  
0                  NaN  
1                  NaN  
2                  NaN  
3                  NaN  
4                  NaN  
5                  NaN  
6                  NaN  
7                  NaN  
8                  NaN  
9                  NaN

In [9]:
# primjetimo da imamo identicna prva dva stupca pa to malo uljepsajmo
test_set.set_index('instance_id', inplace=True)

# pretvorimo i ove datume u datetime
test_set['DATUM_OTVARANJA'] = pd.to_datetime(test_set['DATUM_OTVARANJA'])
test_set['PLANIRANI_DATUM_ZATVARANJA'] = pd.to_datetime(test_set['PLANIRANI_DATUM_ZATVARANJA'])

test_set.head(10)

KLIJENT_ID  OZNAKA_PARTIJE DATUM_OTVARANJA  \
instance_id                                               
0               1049000         5585453      2012-07-14   
1                481302          752335      2016-01-24   
2                838192         6623548      2018-12-26   
3                174575          425060      2012-04-08   
4                 14884         1183149      2011-03-27   
5               1348702        12847410      2018-04-08   
6                764915         5647773      2013-12-21   
7                367476         4875789      2010-08-01   
8                234350        12636299      2017-08-08   
9                888283         9088014      2018-11-23   

            PLANIRANI_DATUM_ZATVARANJA  UGOVORENI_IZNOS  VALUTA  \
instance_id                                                       
0                           2013-07-15        273555.01       2   
1                           2016-01-24         50970.78       2   
2                           2018-12-26          8969.36       2   
3                           2012-08-04         16142.48       2   
4                           2012-03-28         45315.31       2   
5                           2020-02-15         40000.00       1   
6                           2014-03-22        119043.94       2   
7                           2011-01-09        103650.71       2   
8                           2018-08-09        185000.00       1   
9                           2018-11-23         54123.73       1   

             VRSTA_KLIJENTA PROIZVOD VRSTA_PROIZVODA  VISINA_KAMATE  \
instance_id                                                           
0                      1410   TM0109               L           3.60   
1                      1410   TM0109               L           0.20   
2                      1410   TM0109               L           0.02   
3                      1410   TM0109               L           2.30   
4                      1410   TM0109               L           3.30   
5                      1410   TM0109               L           1.50   
6                      1410   TM0109               L           2.00   
7                      1410   TM0109               L           5.15   
8                      1550   TM0109               L           1.30   
9                      1410   TM0109               L           0.20   

            TIP_KAMATE  STAROST  PRIJEVREMENI_RASKID  
instance_id                                           
0                    A       64                  NaN  
1                    A       56                  NaN  
2                    A       52                  NaN  
3                    C       76                  NaN  
4                    C       53                  NaN  
5                    A       35                  NaN  
6                    A       50                  NaN  
7                    A       35                  NaN  
8                    A       46                  NaN  
9                    A       38                  NaN

In [40]:
test_set.shape[0]

84903


### Posto nam je ovaj oblik tablice nepoznat(nema neke feature koje training set ima), uvjerimo se u neke stvari



-----------------------------------------------------------------------------------

* **Posto se orijentiramo na partije(ugovore) provjerimo ima li duplikata medu njima**

In [10]:
o_p = test_set['OZNAKA_PARTIJE']
mask = o_p.isin( o_p[o_p.duplicated()] )

o_p_dt = test_set[mask].sort_values(by=['OZNAKA_PARTIJE'])
o_p_dt.head(25)

KLIJENT_ID  OZNAKA_PARTIJE DATUM_OTVARANJA  \
instance_id                                               
82602            114150          163892      2003-12-02   
82538            114150          163892      2010-07-13   
7914              97642          208868      1999-07-21   
12128             97642          208868      2018-11-17   
8534              97642          208914      2000-04-14   
15991             97642          208914      2018-12-11   
13448             97642          208914      2013-04-21   
10603            210537          209310      2018-08-17   
19899            210537          209310      2001-02-08   
13951            422307          209675      2002-04-22   
1942             422307          209675      2012-07-10   
2051              63623          211234      2007-10-22   
869               63623          211234      2012-04-08   
5244               3702          260126      2018-09-07   
8471               3702          260126      1997-06-17   
12971              3702          260126      2018-10-10   
11638             39569          261263      1998-02-01   
21252             39569          261263      2018-12-15   
7216               4229          262461      1998-09-06   
24241              4229          262461      2018-05-10   
22407             64655          262795      2018-12-28   
14545             64655          262795      1998-07-13   
5637              53743          264363      1998-12-17   
15434             53743          264363      2018-03-12   
9258              29820          267373      2018-01-09   

            PLANIRANI_DATUM_ZATVARANJA  UGOVORENI_IZNOS  VALUTA  \
instance_id                                                       
82602                       2012-12-17      18408597.60       2   
82538                       2016-07-18      18006108.00       2   
7914                        2011-01-17         30628.07       1   
12128                       2018-11-17         35503.91       1   
8534                        2011-01-12         16791.92       1   
15991                       2018-11-12         19472.65       1   
13448                       2013-04-21         18062.78       1   
10603                       2018-08-17         22739.96       1   
19899                       2011-08-10         18042.33       1   
13951                       2011-04-04         39341.21       1   
1942                        2012-10-07         42457.51       1   
2051                        2011-01-29         57914.56       1   
869                         2012-08-04         60861.38       1   
5244                        2018-07-09             0.00       2   
8471                        2011-03-11             0.00       2   
12971                       2018-10-10             0.00       2   
11638                       2011-01-11             0.00       2   
21252                       2018-10-13             0.00       2   
7216                        2011-03-05             0.00       2   
24241                       2018-10-05             0.00       2   
22407                       2018-12-28             0.00       2   
14545                       2011-06-13             0.00       2   
5637                        2011-05-19             0.00       2   
15434                       2018-12-03             0.00       2   
9258                        2018-09-01         13098.78       2   

             VRSTA_KLIJENTA PROIZVOD VRSTA_PROIZVODA  VISINA_KAMATE  \
instance_id                                                           
82602                  1110   FL1310               A           0.00   
82538                  1110   FL1310               A           0.00   
7914                   1410   TM0109               L           3.60   
12128                  1410   TM0109               L           0.10   
8534                   1410   TM0109               L           3.60   
15991                  1410   TM0109               L           0.10   
13448                  1410   TM0109     

In [11]:
o_p_dt.shape[0], o_p[o_p.duplicated()].shape[0]

(5845, 3370)

In [32]:
# pogledajmo kako izgledaju i koliko ima partija sa UGOVORENI_IZNOS = 0
test_set[test_set['UGOVORENI_IZNOS'] == 0.00].sort_values(by='OZNAKA_PARTIJE').head(10)

KLIJENT_ID  OZNAKA_PARTIJE DATUM_OTVARANJA  \
instance_id                                               
12340              1671          256665      2016-03-28   
14339              1588          256833      1996-04-18   
14898              3089          257162      1996-06-26   
5244               3702          260126      2018-09-07   
8471               3702          260126      1997-06-17   
12971              3702          260126      2018-10-10   
16453              8673          260357      1997-07-16   
21252             39569          261263      2018-12-15   
11638             39569          261263      1998-02-01   
10884             48229          261636      2013-07-03   

            PLANIRANI_DATUM_ZATVARANJA  UGOVORENI_IZNOS  VALUTA  \
instance_id                                                       
12340                       2017-03-29              0.0       2   
14339                       2011-01-04              0.0       2   
14898                       2011-07-08              0.0       2   
5244                        2018-07-09              0.0       2   
8471                        2011-03-11              0.0       2   
12971                       2018-10-10              0.0       2   
16453                       2011-01-10              0.0       2   
21252                       2018-10-13              0.0       2   
11638                       2011-01-11              0.0       2   
10884                       2016-03-08              0.0       2   

             VRSTA_KLIJENTA PROIZVOD VRSTA_PROIZVODA  VISINA_KAMATE  \
instance_id                                                           
12340                  1410   TM0109               L           0.70   
14339                  1410   TM0109               L           2.35   
14898                  1410   TM0109               L           3.00   
5244                   1410   TM0109               L           0.02   
8471                   1410   TM0109               L           2.80   
12971                  1410   TM0109               L           0.02   
16453                  1410   TM0109               L           2.20   
21252                  1410   TM0109               L           0.02   
11638                  1410   TM0109               L           2.70   
10884                  1410   TM0109               L           2.55   

            TIP_KAMATE  STAROST  PRIJEVREMENI_RASKID  
instance_id                                           
12340                A       77                  NaN  
14339                C       61                  NaN  
14898                C      900                  NaN  
5244                 A       80                  NaN  
8471                 C       73                  NaN  
12971                A       80                  NaN  
16453                C       61                  NaN  
21252                A       68                  NaN  
11638                C       60                  NaN  
10884                C       18                  NaN

In [33]:
test_set[test_set['UGOVORENI_IZNOS'] == 0.00].sort_values(by='OZNAKA_PARTIJE').shape[0]

42

In [31]:
# pogledajmo kako izgledaju i koliko ima partija sa kasnijim datumom otvaranje nego pl. datuma zatvaranja
test_set[(test_set['PLANIRANI_DATUM_ZATVARANJA'] - test_set['DATUM_OTVARANJA']) < pd.Timedelta(days=0)].sort_values(by='OZNAKA_PARTIJE').head(10)

KLIJENT_ID  OZNAKA_PARTIJE DATUM_OTVARANJA  \
instance_id                                               
15991             97642          208914      2018-12-11   
5244               3702          260126      2018-09-07   
5016               8362          260556      2012-12-01   
21252             39569          261263      2018-12-15   
23809             83931          265671      2015-11-03   
1449             175007          272480      2014-09-02   
81427              3327          277351      2015-11-06   
81323              3327          277351      2017-12-06   
5720             314845          279464      2018-12-10   
24803            238289          280419      2018-12-27   

            PLANIRANI_DATUM_ZATVARANJA  UGOVORENI_IZNOS  VALUTA  \
instance_id                                                       
15991                       2018-11-12         19472.65       1   
5244                        2018-07-09             0.00       2   
5016                        2012-07-13         66829.76       3   
21252                       2018-10-13             0.00       2   
23809                       2015-09-12         93392.04       2   
1449                        2014-02-09         59513.62       2   
81427                       2015-06-11         11881.09       2   
81323                       2017-06-12         12231.25       2   
5720                        2018-10-12          5913.23       2   
24803                       2018-10-25         69961.11       2   

             VRSTA_KLIJENTA PROIZVOD VRSTA_PROIZVODA  VISINA_KAMATE  \
instance_id                                                           
15991                  1410   TM0109               L           0.10   
5244                   1410   TM0109               L           0.02   
5016                   1410   TM0109               L           2.40   
21252                  1410   TM0109               L           0.02   
23809                  1410   TM0109               L           1.45   
1449                   1410   TM0109               L           2.10   
81427                  1410   FD0100               L           1.70   
81323                  1410   FD0100               L           0.30   
5720                   1410   TM0109               L           0.05   
24803                  1410   TM0109               L           0.02   

            TIP_KAMATE  STAROST  PRIJEVREMENI_RASKID  
instance_id                                           
15991                A       84                  NaN  
5244                 A       80                  NaN  
5016                 C       48                  NaN  
21252                A       68                  NaN  
23809                A       91                  NaN  
1449                 A       40                  NaN  
81427                A       48                  NaN  
81323                A       50                  NaN  
5720                 A       17                  NaN  
24803                A       77                  NaN

In [30]:
test_set[(test_set['PLANIRANI_DATUM_ZATVARANJA'] - test_set['DATUM_OTVARANJA']) < pd.Timedelta(days=0)].shape[0]

2252

Primjecujemo da postoje duplikati kojih ima 3370, ali mozemo i uociti da se ipak ne radi o istim redovima u tablici. Naime, taj ugovor se produzavao(pogledati objasnjenje u FAQ-u) te bi takve ugovore **trebali** "spojiti" u jedan. Plus za napraviti to je i taj da vecina duplikata(od ovih prvih 25) ima isti datum otvaranja i planirani datum zatvaranja sto vjerojatno znaci da se radi o nekoj bankarskoj formalnosti radi koje se treba raditi novi (produzeni) ugovor te zbog toga nema smisla taj duplikat shvacati kao zaseban ugovor.

Druga stvar je da moramo izbaciti partije ciji je UGOVORENI_IZNOS 0 (pogledati objasnjenje u FAQ-u). 99% sam siguran da nam te partije ne trebaju, a i nema ih puno(42).

U biljeske.md stavljam i neke ostale probleme koje cemo morati rijesiti na neki drugi nacin(npr. slati mail)

-----------------------------------------------------------------------------------

* **Provjerimo sada pojavljuju li se ponovno cudne vrijednosti za STAROST**

In [39]:
# pogledajmo koliko ima takvih cudnih vrijednosti
test_set[np.logical_or(test_set['STAROST'] < 0, test_set['STAROST'] > 500)].shape[0]

154

Ovo zapravo znaci da necemo moci samo tako izbaciti redove ciju starost ne znamo(pogledati FAQ) jer ocito dolaze i u ovom test skupu(pa i u validacijskom kojeg necemo vidjeti sve do pred kraj) te ce nas model morati znati handle-ati ovakve stvari...

Rjesenje ovog problema bi mogli rijesiti mean/median solucijom ili mozda da popunimo te cudne vrijednosti sa nekim vrijednostima blizu mean-a(na osnovu std-a tog stupca STAROST) ili jednostavno bubnemo neku konstantu(po meni manju od nule) koja ce oznacavati tu ekipu kojoj se ne zna starost jer mozda su ti likovi nekako povezani. Izvor ovih ideja je [ovaj](https://towardsdatascience.com/handling-missing-values-in-machine-learning-part-1-dda69d4f88ca)